### DISCUSSION TITLE
Download folder as zip

### DISCUSSION THREAD

**Discussion by felix2056** [qgustavor/mega#115] [Q&A] [answered]
Can you write a code snippet that zips a folder (My Directory) containing multiple mp4 files and then downloads the zipped file as my-directory.zip?

I tried this but didn't get the best results

```
file.on('data', (data) => {
        // zip file
        const blob = new Blob([data], { type: 'application/zip' })
        const url = window.URL.createObjectURL(blob)

        // download zip file
        const a = document.createElement('a')
        a.href = url
        a.download = folder.name
        a.click()
});
```

**Comment by qgustavor** [comment, marked as answer]
Look, this is not StackOverflow, but I guess what you're trying to implement is something like this:

```javascript
import { File } from 'https://cdn.skypack.dev/megajs'
import JSZip from 'https://cdn.skypack.dev/jszip'
import saveAs from 'https://cdn.skypack.dev/file-saver'

const folder = File.fromURL('link to some folder') // you can also load it from your account
await folder.loadAttributes() // this is not needed if you load from your account

const handleFolder = async (megaFolder, zipFolder) => {
  if (!megaFolder.children) return
  for (const file of megaFolder.children) {
    if (file.directory) {
      await handleFolder(file, zip.folder(file.name))
    } else {
      zip.file(file.name, await file.downloadBuffer())
    }
  }
}

const zip = new JSZip()
await handleFolder(folder, zip)

saveAs(await zip.generateAsync({type: 'blob'}), 'folder.zip')
```

I didn't tested this code, but should from my experience using those three libraries it might work. Maybe. I don't know. Well most of this question is not related to the library but JavaScript related so that's better suited to StackOverflow. Here are the keypoints of the above example (some related to the library):

- Nodes loaded from `File.fromURL` or loaded from the `Storage` class can be files or folders... even if the class is named `File`. Hey, don't blame me, blame backyards compatibility.
- Use async/await. It is one the best features of megajs@1. Calling `.on('data', ...)` will just return the first downloaded chunk and you will download an incomplete file.
- There is a tutorial for downloading files - https://mega.js.org/docs/tutorial/downloading Is the tutorial unclear somehow? If so I will fix it.
- The library does not provide helper functions - like some function for recursing folders - because they often end being too easy to implement - well, at least for me it's easy - and might ending bloating the library which is not good since it's meant to be used also in browsers, not just Node.js or Deno. Maybe v2 will have some helper functions as the plan is to make it more tree shakable than it's now. Maybe a section for "Handling folders" with some examples of recursive functions would be a nice fit to the tutorial.
- Want to zip files? There is a library for that: https://stuk.github.io/jszip/ (maybe there are other libraries, I'm just used to this one)
- Want to download files? There is a library for that also: https://www.npmjs.com/package/file-saver (doing the a.download think might work in some browsers, in the other hand the library handles edge cases)
- Please don't ignore the issue template that says to not open an issue for things like that and open a discussion instead. Better, ask on StackOverflow next time. I wrote all of this because [I'm used to this kind of task](https://github.com/qgustavor/mkv-extract/blob/850b8a74bb24145d692df7ee95578488c9684a85/src/main.ts#L76-L94) so I just moved it to a discussion.

**Reply by felix2056** [reply]
Thanks, @qgustavor. That worked
One last thing if you don't mind.
Can you tell how to call on progress event on each of the files being downloaded?

**Reply by felix2056** [reply]
Nevermind i just read the doc

**Reply by qgustavor** [reply]
Some example code for that: https://github.com/qgustavor/direct-mega/blob/8897b53f2e547dd0948cb4923a7c1ed51ebaf3b6/src/fallback.js#L121-L135
